62-california-condor/technical_specification.md

```diff
@@ -0,0 +1,158 @@
+  }
+]
+```
+
+The migration process will form a chain, where discreet migration logic exists for
+every database version to migrate the data from version X to X+1. Most of the time, the
+database will be current or, at most, one version behind. However, if a database restore
+occurs and the data happens to be older, the migration process will begin at the
+appropriate step in the migration chain and continue until the data is totally migrated.
+Migration code should be preserved at least until there is no possibility of 
+encountering the corresponding database version again.
+
+
+### Services with Multiple Entrypoints
+
+Several services operate with more than one instance simultaneously because one serves
+as a REST API and another consumes Kafka events (for example). Obviously, only one
+migration process should occur in these situations, rather than executing for each
+instance. This can be solved if we ensure the migration process only runs as part of
+the startup for one entrypoint, e.g. the REST API. We can "lock" the database to signal
+to any other potential instances that there is already a migration in progress:
+
+```json
+// locking collection with one document
+[
+  {
+    "migration_in_progress": true
+  }
+]
+```
+
+The first instance to obtain the lock is allowed to proceed, all others wait. This has
+the benefit of preventing concurrent migrations if services are scaled. An alternative
+to a dedicated collection is to use the database version collection -- if the
+expected database version exists but the timestamp is missing, then that means a
+migration is already underway.
+
+After preventing simultaneous migrations, we need to optimize how other instances
+operate while a migration is underway and consider how to handle read and 
+write requests. One way would be to accept some downtime and
+block service instances until the migration is complete. This might be a
+sufficient initial solution because database sizes are small enough that migrations will
+be completed quickly. If we have to eliminate downtime, we could perform shadow
+migrations and write the results to temporary collections while the old service
+version continues to handle requests. When the migration is complete, the old service
+can be taken offline and the collections swapped out. That's a little more complex.
+
+### Reverse Migrations
+
+There might be some situation where we need to apply the reverse of a migration.
+If DB version 5 is applied, but we later find that we should have stayed with version 4,
+then we need to move to version 6. Version 6 is not treated as some special 'undo'
+version increment; the migration logic merely happens to move the data to the same
+structure contained in version 4.
+
+### Migration Structure
+
+![Migration structure](./images/db%20migrations%20white%20bg.png)
+
+In the image above, the green section indicates the top-level migration logic, most of
+which can be implemented in a library like `hexkit`.  
```

Comment:
Where it says "get docs" did you mean "get collections"?

The diagram should also show what happens if there is an error because otherwise it's not evident why we need temporary collections.

Reply:
I like the idea of having a framework to handle the most frequent case of updating every document in one or more collections to a new schema version. But the framework should also provide a hook to execute more complex custom migrations, like merging two collections into one.